### PR TITLE
Remove invalid nodes

### PR DESCRIPTION
**Pull request by terencechain**
Make sure we can remove invalid blocks and states up to `latest_valid_hash`

Change list:
- Add `removeInvalidBlockAndState` and gets called when `ErrInvalidPayloadStatus` is responded from `NewPayload`
- DB returns `ErrDeleteJustifiedAndFinalized` to signal an attempt to remove justified and finalized state in DB. Caller can decide to do panic or other things
- DB allows deletion of head state. This is accepted and required because head could be optimistic and could become invalid
- DB when delete block also deletes state summary
- Add `DeleteStateFromCaches` setters to state gen service

Blocked by:
~~- Insert payload hash to forkchoice store, this will enable very fast look up~~
~~- Clarification on whether we should panic with an invalid justified and finalized state~~
~~- #10398~~
~~- #10403~~

### CODE REVIEW

**Review by potuz — COMMENTED** [review]
I think the deletion is a bit more complicated: one needs to go up to latest valid hash and then descend back to every child.

File: beacon-chain/blockchain/optimistic_sync.go [hunk 1]
```diff
@@ -196,3 +203,21 @@ func (s *Service) optimisticCandidateBlock(ctx context.Context, blk block.Beacon
...
+		if err := s.cfg.StateGen.DeleteStateFromCaches(ctx, root); err != nil {
+			return err
+		}
+
+		// Delete block also deletes the state as well.
+		if err := s.cfg.BeaconDB.DeleteBlock(ctx, root); err != nil {
+			if err == kv.ErrDeleteJustifiedAndFinalized {
+				log.Panic("Invalid justified / finalized block in DB. Please resync from last weak subjectivity checkpoint")
+			}
+			return err
```
[the remark this answers is not in the document]
Comment (terencechain) [reply beacon-chain/blockchain/optimistic_sync.go:201]: No it doesn't, I'll add a unit test for this

File: beacon-chain/blockchain/optimistic_sync.go [hunk 2]
```diff
@@ -123,6 +124,19 @@ func (s *Service) notifyNewPayload(ctx context.Context, preStateVersion, postSta
 				"blockHash": fmt.Sprintf("%#x", bytesutil.Trunc(payload.BlockHash)),
 			}).Info("Called new payload with optimistic block")
 			return false, nil
+		case powchain.ErrInvalidPayloadStatus:
+			root, err := blk.Block().HashTreeRoot()
+			if err != nil {
+				return false, err
+			}
+			invalidRoots, err := s.ForkChoicer().SetOptimisticToInvalid(ctx, root, bytesutil.ToBytes32(payload.BlockHash))
```
[the remark this answers is not in the document]
Comment (terencechain) [reply beacon-chain/blockchain/optimistic_sync.go:132]: Fixed in [a010026](https://github.com/prysmaticlabs/prysm/pull/10399/commits/a010026f16b0674e005030c09c6a910ef95c35f9)

File: beacon-chain/blockchain/optimistic_sync.go [hunk 3]
```diff
@@ -174,3 +188,21 @@ func (s *Service) optimisticCandidateBlock(ctx context.Context, blk block.Beacon
...
+	for _, root := range blkRoots {
+		if err := s.cfg.StateGen.DeleteStateFromCaches(ctx, root); err != nil {
+			return err
+		}
+
+		// Delete block also deletes the state as well.
+		if err := s.cfg.BeaconDB.DeleteBlock(ctx, root); err != nil {
+			if err == kv.ErrDeleteJustifiedAndFinalized {
+				log.Panic("Invalid justified / finalized block in DB. Please resync from last weak subjectivity checkpoint")
+			}
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/optimistic_sync.go:203]: I don't think a panic is the right choice here. Can you surface the error further up the application?

Comment (terencechain) [reply beacon-chain/blockchain/optimistic_sync.go:203]: If this error happens, the node should be stopped. Are you suggesting surfacing this specific error further up the application to panic or to gracefully shut down the node?

**Review by potuz — APPROVED** [review]

**Review by potuz — APPROVED** [review]

**Pull request merged**